sysl2/sysl/importer/grammar.go

```diff
@@ -0,0 +1,190 @@
+package importer
+
+import (
+	"bytes"
+	"fmt"
+
+	parser "github.com/anz-bank/sysl/sysl2/naive"
+	ebnfGrammar "github.com/anz-bank/sysl/sysl2/proto"
+	"github.com/sirupsen/logrus"
+)
+
+type grammarParam struct {
+	grammarInput string
+	grammarName  string
+}
+
+type Grammar struct {
+	grammarParam
+	grammar    *ebnfGrammar.Grammar
+	logger     *logrus.Logger
+	types      TypeList
+	visitedMap map[string]bool
+}
+
+func LoadGrammar(args OutputData, grammarInput string, logger *logrus.Logger) (out string, err error) {
+	gParam := grammarParam{
+		grammarInput: grammarInput,
+		grammarName:  args.AppName,
+	}
+	return newGrammar(gParam, logger).writeSysl()
+}
+
+func newGrammar(param grammarParam, logger *logrus.Logger) *Grammar {
+	return &Grammar{
+		grammarParam: param,
+		logger:       logger,
+		types:        TypeList{},
+		visitedMap:   map[string]bool{},
+	}
+}
+
+func (g *Grammar) createChoice(choice *ebnfGrammar.Choice, name string) Type {
+	choiceType := Union{
+		name: name,
+	}
+	// mark the node as visited if not visited before
+	if !g.visitedMap[name] {
+		g.visitedMap[name] = true
+	}
+	for _, seq := range choice.Sequence {
+		for _, term := range seq.Term {
+			switch x := term.Atom.Union.(type) {
+			case *ebnfGrammar.Atom_String_:
+				// hit the leaf node
+				// return
+			case *ebnfGrammar.Atom_Choices:
+				choiceType.Attributes = append(choiceType.Attributes, "__Choice_"+name)
+
+				// visit the unvisited node
+				if !g.visitedMap["__Choice_"+name] {
+					childChoiceType := g.createChoice(x.Choices, "__Choice_"+name)
+					g.types.types = append([]Type{childChoiceType}, g.types.types...)
+				}
+			case *ebnfGrammar.Atom_Rulename:
+				var childRuleType Type
+				choiceType.Attributes = append(choiceType.Attributes, x.Rulename.GetName())
+
+				// visit the unvisited node
+				if !g.visitedMap[x.Rulename.GetName()] {
+					if rule, exists := g.grammar.Rules[x.Rulename.GetName()]; exists {
+						childRuleType = g.createRule(g.grammar.Rules[rule.Name.Name], rule.Name.Name)
+					} else {
+						childRuleType = g.createEnumType(x.Rulename.GetName())
+					}
+					g.types.types = append([]Type{childRuleType}, g.types.types...)
+				}
+			default:
+				g.logger.Warnf("unexpected term type")
+			}
+		}
+	}
+	var baseType Type = &choiceType
+	return baseType
+}
+
+func (g *Grammar) quantifyUnionType(fieldName, fieldType string, quant *ebnfGrammar.Quantifier) Field {
+	if quant == nil {
+		return Field{
+			Name: fieldName,
+			Type: &StandardType{
+				name: fieldType,
+			}}
+	}
+	switch quant.Union.(type) {
+	case *ebnfGrammar.Quantifier_Optional:
+		fieldType = fmt.Sprintf("%s?", fieldType)
+	case *ebnfGrammar.Quantifier_ZeroPlus:
+		fieldType = fmt.Sprintf("sequence of %s?", fieldType)
+	case *ebnfGrammar.Quantifier_OnePlus:
+		fieldName = fmt.Sprintf("%s(1..)", fieldName)
+	default:
+		g.logger.Warnf("unexpected quantifier type")
```

Comment:
This should probably be a fatal error, not a warning

Reply:
It already caters to all the know types as of now, tomorrow if the protobuff changes, I don't want the conversion to fail but let it proceed with a warning.

Reply:
tomorrow if the protobuf changes we 100% want to make sure the person who changes it fixes the importer to deal with it. Make this fatal